src/lib.rs

```diff
@@ -1,215 +1,64 @@
-extern crate futures;
-extern crate parking_lot;
-
-use parking_lot::Mutex;
-use futures::prelude::*;
-use futures::task::{self, Task, AtomicTask};
-use futures::executor::{self, Notify};
-
-use std::collections::HashMap;
 use std::sync::Arc;
-use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+use std::pin::Pin;
+use std::task::{Poll, Context};
+use std::ops::DerefMut;
+use futures::{Future, FutureExt, channel::oneshot, future::{select, Either}, executor::block_on};
+use parking_lot::Mutex;
 
-/// Future that resolves when inner work finishes or on exit signal firing.
+/// Future that resolves when the exit signal has fired.
 #[derive(Clone)]
-pub struct UntilExit<F> {
-    inner: F,
-    exit: Exit,
-}
-
-impl<F: Future> Future for UntilExit<F> {
-    type Item = Option<F::Item>;
-    type Error = F::Error;
-
-    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        match self.inner.poll() {
-            Ok(Async::Ready(x)) => Ok(Async::Ready(Some(x))),
-            Ok(Async::NotReady) => Ok(self.exit.check().map(|()| None)),
-            Err(e) => Err(e),
-        }
-    }
-}
-
-struct Notifier {
-    signalled: AtomicBool,
-    outer_task: AtomicTask,
-}
+pub struct Exit(Arc<Mutex<oneshot::Receiver<()>>>);
 
-impl Notify for Notifier {
-    // id isn't necessary because this is a notifier for a single
-    // exit future.
-    fn notify(&self, _: usize) {
-        self.signalled.store(true, Ordering::Release);
-        self.outer_task.notify()
-    }
-}
-
-struct ExitInner {
-    shared_id: usize,
-    notifier: Arc<Notifier>,
-}
-
-impl ExitInner {
-    fn check(&self, shared: &Shared) -> Async<()> {
-        // set up the outer task correctly so when the inner
-        // task is notified, we get woken up.
-        self.notifier.outer_task.register();
-
-        // if any poll was signalled by the inner task, use it.
-        if !self.notifier.signalled.fetch_and(false, Ordering::SeqCst) {
-            return Async::NotReady
-        }
+impl Future for Exit {
+    type Output = ();
 
-        // do heavy check with inner task. ID doesn't matter.
-        executor::with_notify(&self.notifier, 0, || {
-            if shared.is_live_and_notify(self.shared_id) {
-                Async::NotReady
-            } else {
-                Async::Ready(())
-            }
-        })
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let mut receiver = Pin::into_inner(self).0.lock();
+        Pin::new(receiver.deref_mut()).poll(cx).map(|_| ())
     }
 }
 
-/// Future that resolves when the exit signal has fired.
-pub struct Exit {
-    inner: Option<ExitInner>,
-    shared: Arc<Shared>,
-}
-
 impl Exit {
-    /// Check if the signal is live outside of the context of a task and 
-    /// without scheduling a wakeup.
+    /// Check if the signal is live (hasn't been cancelled).
     pub fn is_live(&self) -> bool {
-        self.shared.waiting.lock().0
+        self.0.lock().try_recv().is_ok()
```

Comment:
should rather be `is_err`? If there is something in the channel we are not live any more, no? Also if you consume the response here, polling would fail.

Reply:
I changed it to check for `Ok(None)`: yes the exit signal hasn't been 'cancelled' (dropped), no we haven't recieved anything yet. This is more clear.

Reply:
I think it's still subject to race - it will fail in case of `Ok(Some(_))`, perhaps `is_ok()` was better. I'm still curious though if it won't break `poll` after we get `Ok(Some(_))` here, but that's the same issues as mentioned earlier (i.e. `poll` again after receiving a result).

Reply:
I only kept `is_live` here because it was in the previous version. It's not used in either substrate or polkadot, and we can probably remove it.